### ISSUE TITLE
[regression/8.0.3] `TapGestureRecognizer` no longer works on `Button`

### ISSUE THREAD

**Issue by Ghostbird** [dotnet/maui#19099]
### Description

We updated our app to MAUI 8 and built in on Android, and everything worked fine. However we were shocked when we started testing the iOS app. Almost none of the buttons were working.

After more testing we concluded that any button behaviour that relied on a `TapGestureRecognizer` didn't work any more. Even if we manually bound the `Clicked` event instead, it would not work if we also had e.g. `AnimationBehavior` (which uses a `TapGestureRecognizer` under the hood by default) attached.

I looked through the code, and I can't see any recent change to the iOS `Button` code or the `TapGestureRecognizer` code that could have caused this change, or I'd have tried to PR a fix myself.

The only buttons that kept working were ones that bound `Clicked` in XAML, but as our app contains only a single static settings page, and the rest of the content is dynamically loaded and rendered, the MAUI 8 version of app is completely unusable on iOS.

### Steps to Reproduce

1. Create a new MAUI 8 app.
2. Add a new page e.g. `ButtonTestPage.cs`:

    ```csharp
    public class ButtonTestPage : ContentPage
    {
      public ButtonTestPage()
      {
        var button = new Button();
        button.GestureRecognizers.Add(new TapGestureRecognizer() {
          Command = new Command(() => DisplayAlert("Yay!", "It works!", "Ok"))
        });
        Content = button;
      }
    }
    ```

3. Ensure that page is initially loaded.
4. Build an Android version and an iOS version.
5. Test both:
    - Button works fine in Android
    - Button doesn't work in iOS.

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

7.0.101

### Affected platforms

iOS

### Affected platform versions

Observed on iOS 15, 16, 17, potentially affects all.

### Did you find any workaround?

For now we've made two extension methods, that we use whenever we need to add tap handling to any element. The code below is not exactly what we use, as we wrote something that better integrates with `CommunityToolkit.Maui.Markup`'s fluent API, and adds extra behaviours for graphical and haptic feedback. However, the essence of the workaround is that we leverage extension method resolution order to use an `EventToCommandBehavior` instead of a `TapGestureRecognizer` specifically for `Button`s. The big advantage is that when this issue gets fixed, we only need to remove the `Button` specific extension method, no other code changes are required:

```csharp
public static class TapExtensions
{
  public static void AddTapGesture<TButton>(this TButton self, Action onTapped)
    where TButton : Button
  {
    self.Behaviors.Add(new EventToCommandBehavior
        {
          EventName = nameof(Button.Clicked),
          Command = new Command(onTapped)
        })
  }
  
  public static void AddTapGesture<TView>(this TView self, Action onTapped)
    where TView : View, IGestureRecognizers
  {
    self.GestureRecognizers.Add(new TapGestureRecognizer() {
      Command = onTapped
    });
  }
}
```

**Comment by Ghostbird** [comment]
Yes, I understand that, and it makes perfect sense when you're building static pages in XAML. However we dynamically create the controls based on configuration. In our situation, it's much more elegant to call an extension method works on any element (which has worked nicely in MAUI up to now) that is configured to handle a tap event. We used to use `TapGesture` from `CommunityToolkit.Maui.Markup`. Now we use something similar to the workaround I describe.

Even then, we think it's quite surprising that this big of a change apparently went completely unnoticed during the MAUI 8 release.

**Comment by drasticactions** [comment]
But then you wouldn't need to use `Button`, you could write your View and layout and handle that with a TapGestureRecognizer instead. Why do you need to use a Button?

**Comment by Engisan** [comment]
I face issues with GestureRecognizers on other controls as well. It depends on where they are placed in the UI. But there must be some problem with them globally. :/

**Comment by Ghostbird** [comment]
This getting off-topic. Why would we use MAUI and then write all the controls ourselves? That's antithetical to library use. The whole reason we were hit by this issue is that we want to use as _little_ custom code as possible. That's not fragile, using a well-defined library API is common sense. It's much _less_ likely to break than anything we can cook up ourselves.

Of course it's not XAML specific. To me your argument is analogous to this:

I'm used to driving all kinds of vehicles that have an option to `Reverse()`. But now the reverse of my boat is broken.

>‌ Why would you use the `Reverse()`? That type of boat has a paddle under the seat that you can use to go backwards. Obviously the reverse gear should be fixed, but I don't understand why you're using it in the first place.

When I put it that extremely, it becomes obviously weird. However it illustrates that I prefer to use generic API functionality over highly implementation specific code. It provides better decoupling of abstraction layers and separation of concerns, and uses less code to achieve the same thing. Using a highly specific implementation should only be done when there's a highly specific _need_ to do so (such as a workaround for this bug), and even then, it should be abstracted away so the rest of the application code is not affected in any way (hence the workaround design that leverages compile-time extension method resolution).

**Comment by drasticactions** [comment]
I apologize, it seemed to me like there were other ways to accomplish what you were trying to do that may have made more sense, but I understand what you're saying. Again, not saying this wasn't an issue, was just trying to help.

**Comment by Ghostbird** [comment]
No, I get that, an you point is valid. Now we use those other ways as a workaround, but we preferred using the most broadly supported API, which turned out to be `TapGestureRecognizer`.

**Comment by ryansmithsquaresix** [comment]
I think a Button should work with Command and not TapGestureRecognizer. We also had a Helper that used to add TapGestureRecognizer to Controls dynamically (probably similar to your logic), and that did break in .NET 8. But we check if the Control has Command property first now and add the Command if it's there, otherwise default to TapGestureRecognizer.

**Comment by crs2m5smith** [comment]
I think a Button should work with Command and not TapGestureRecognizer. We also had a Helper that used to add TapGestureRecognizer to Controls dynamically (probably similar to your logic), and that did break in .NET 8. But we check if the Control has Command property first now and add the Command if it's there, otherwise default to TapGestureRecognizer.

**Comment by ryansmithsquaresix** [comment]
I think it is a pretty clean and improved work around. This fix could also help from future breaks again from maui. I understand we always want the framework to do the heavy lifting, but also our code should be resilient to changes as well (which happen often with maui).

Im just saying we simply check for command property on control first. And then if it doesn't have one we add it as tap gesture and our controls are working again. Waiting for a fix from maui team for this issue likely wont happen for a while unfortunately. 

It seems the fix would be for any control that has a command, it should check tap gesture first. If null then try command property. And that fix can either be in the framework or in our code base

**Comment by Ghostbird** [comment]
Don't you think that's kind of weird? Almost all MAUI controls are based on `View`, which implements `IGestureRecognizers`. You're proposing we do either of these:

- Rewrite MAUI so each subclass of `View` can individually implement `IGestureRecognizers`.
- Implement `GestureRecognizers => throw new NotSupportedException()` on at least the `Button`.

Instead we could simply treat this like a bug, fix it, and being left with a framework that is neither of these:

- Was heavily rewritten to match a reality that could have been treated as a single regression.
- In violation of the definitions of the language it's written in.

**Comment by Ghostbird** [comment]
> @mattleibow Could you elaborate why this was triaged as `Backlog` rather than something else? A bug that prevents all buttons from working on a specific platform for those who consume the generic API, rather than the `Button` specific API, seems fairly high impact to me.
> 
> Especially as those who implicitly consume that API are also affected. You have a `Button` that uses `Clicked` but you also added an `AnimationBehavior`? Button clicks won't work, because `AnimationBehavior` has added a `TapGestureRecognizer`.

@mattleibow Could you comment on the choice of prioritisation of this issue? It seems like a fairly serious regression to me.

**Comment by PureWeen** [comment]
This most likely regressed with this PR https://github.com/dotnet/maui/pull/12108 as the gestures are probably being added to the containerview behind the button

**Comment by Ghostbird** [comment]
In that case, could I verify that by setting the button `InputTransparent`? I'd think that then the tap would be handled by the containerview below.

**Comment by mnxamdev** [comment]
Can confirm updating to use .Net 8 breaks button tapping on iOS. Additionally it breaks text input tapping. Do we have an ETA on a release for the fix?

**Comment by sisaacks** [comment]
I am seeing an issue with entry tap gesture inside of data template.  The event is not fired on iOS but it is on Android.

**Comment by RhomGit** [comment]
I just wanted to add that I also am experiencing the same symptom, albeit with Grid not Button.
```
                    <DataTemplate>
                        <Grid ColumnDefinitions="Auto,*,Auto,Auto">
                            <Grid.GestureRecognizers>
                                <TapGestureRecognizer Tapped="Node_Tapped" />
                            </Grid.GestureRecognizers>
```
NET9, the event fires on Android, not on iOS.

**Comment by marcojak** [comment]
Not sure if it's the same issue but the tapgesturerecogniser doesn't work for me in most of the cases both on android and iOS.

I'll try to investigate further but right now they are unusable

**Comment by Barry-Theunissen** [comment]
I have the same issue for an IMAGE BUTTON tap gesture recognizer not working on iOS, but works fine on Android. A seriously breaking change this, and must be fixed!

Example code which refuses to bind to the "MenuCommand" correctly:

            <ImageButton AutomationId="ContextMenu" Grid.Column="4" Source="hamburgericon.png" HorizontalOptions="End">


                <ImageButton.GestureRecognizers>
                    <TapGestureRecognizer Command="{Binding MenuCommand}" />
                </ImageButton.GestureRecognizers>
            </ImageButton>

**Comment by sisaacks** [comment]
Is the image button in a datatemplate?
Also why are you not using a click event?
I think maui has a click to command converter...something like that.

**Comment by kmoroz** [comment]
I'm experiencing the same issue on iOS (works fine on Android) ⚰️

```
<Label>
    <Label.FormattedText>
        <FormattedString>
            <Span>
	        <Span.GestureRecognizers>
	            <TapGestureRecognizer Tapped="OnTapped" />
	        </Span.GestureRecognizers>
	    </Span>
	</FormattedString>
    </Label.FormattedText>
</Label>
```

**Comment by bhavanesh2001** [comment]
@kmoroz  could you please raise a new bug with a simple repro

**Comment by vnobo** [comment]
`  <HorizontalStackLayout 
      Spacing="15" Padding="30,0"
      BindableLayout.ItemsSource="{Binding Projects}">
      <BindableLayout.ItemTemplate>
          <DataTemplate x:DataType="models:Project">
              <controls:ProjectCardView WidthRequest="200">
                  <controls:ProjectCardView.GestureRecognizers>
                      <TapGestureRecognizer Command="{Binding NavigateToProjectCommand, Source={RelativeSource AncestorType={x:Type pageModels:MainPageModel}}, x:DataType=pageModels:MainPageModel}" CommandParameter="{Binding .}"/>
                  </controls:ProjectCardView.GestureRecognizers>
              </controls:ProjectCardView>
          </DataTemplate>
      </BindableLayout.ItemTemplate>
  </HorizontalStackLayout>`

**Comment by bnSonic** [comment]
I have this serious problem on iOS too (works on Android) - TapGesture of a Grid or StackView doesn't trigger when inside a DataTemplate used in a CollectionView.
In my case a bit more special - so i have built a stipped down demo and opened a new issue for that:
[https://github.com/dotnet/maui/issues/29880](url)

Because this problem is a stopper for the iOS version of the MAUI App i really need some kind of solution for this :( 
I've tried so many things for two days now until i was able to stip that problem down to this simple demo-app - but until now i wasn't able to fine a fix.
So i hope that someone knows a working work-around?

**Comment by RhomGit** [comment]
For anyone (still) following this, I have tried this answer from Jeremie and it worked:
https://stackoverflow.com/a/79535537/8755531

**Comment by amortensen13** [comment]
I'm glad the workaround worked.  It would be nicer if it hadn't been broken.  Can I vote for more testers?  The programmers are doing great, they just need more support.  I'm happy to talk to the non-programmers over the money that don't think they're needed.
